### ISSUE TITLE
ucx initialization + cudf

### ISSUE THREAD

**Issue by quasiben** [rapidsai/ucx-py#184]
ucx initialization is suppose to happen before any cuda context loading -- @Akshay-Venkatesh has said this several times most recently, [here](https://github.com/rapidsai/ucx-py/pull/180#issuecomment-530957458)

Dask doesn't call `ucp.init()` until a worker has started up and a connection has been made.  I believe this happens *after* [preload modules](https://github.com/dask/distributed/blob/master/distributed/worker.py#L987-L996) haven been loaded.

When using dask-cuda the preload modules are used to [set the cuda context](https://github.com/rapidsai/dask-cuda/blob/branch-0.10/dask_cuda/initialize_context.py).  It is possible, therefore, that dask is not calling `ucp.init` before calling `numba.cuda.current_context()` and this may cause problems.

As a test I patched the dask-cuda preload with `ucp.init()` prior to calling `current_context()` and inserted `ucp.init() inside [ucx.py](https://github.com/dask/distributed/blob/master/distributed/comm/ucx.py#L29.  The tests failed immediately with the following error

```
[1568323224.190797] [dgx15:47683:0]      rdmacm_ep.c:52   UCX  ERROR rdma_create_id() failed: Permission denied
failed to connect to 10.33.225.165 (Input/output error)
Unable to start CUDA Context
Traceback (most recent call last):
  File "/home/nfs/bzaitlen/miniconda3/envs/cudf-dev/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/nfs/bzaitlen/miniconda3/envs/cudf-dev/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/nfs/bzaitlen/miniconda3/envs/cudf-dev/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [3] Call to cuInit results in CUDA_ERROR_NOT_INITIALIZED
```

@kkraus14 @mt-jones and other knowledgable cuda/numba folks, what happens if you create a cuda context in one thread and attempt share it in other threads you spawn ?

@Akshay-Venkatesh @madsbk and other UCX folks, what happens if you call ucp.init() in one thread -- do you need to init in all threads ?

**Comment by quasiben** [comment]
@Akshay-Venkatesh Does UCX or UCX-PY ever change the cuda context ?

**Comment by quasiben** [comment]
@mrocklin does the dask-scheduler ever deserialize a cuda object ?  Would you expect to set the env var `CUDA_VISIBLE_DEVICES` on the scheduler

**Comment by Akshay-Venkatesh** [comment]
No

**Comment by mrocklin** [comment]
The scheduler doesn't deserialize anything.  However, it may proxy things.  So it may still receive GPU data and so we might want to specify CUDA_VISIBLE_DEVICES there in order to target which GPU it uses, although always choosing GPU 0 probably makes sense.

**Comment by pentschev** [comment]
I did a simple test on this, starting 4 processes as follows:

1. `CUDA_VISIBLE_DEVICES="" dask-scheduler`
1. `CUDA_VISIBLE_DEVICES="0" dask-cuda-worker $SCHEDULER_IP:$SCHEDULER_PORT`
1. `CUDA_VISIBLE_DEVICES="1" dask-cuda-worker $SCHEDULER_IP:$SCHEDULER_PORT`
1. `CUDA_VISIBLE_DEVICES="2" python

On the last process I executed:

```python
import cupy, dask, dask.array as da
from distributed import Client
client = Client()  # Scheduler's "IP:PORT" as argument
rs = da.random.RandomState(RandomState=cupy.random.RandomState)
rs.normal(10, 1, (10000, 10000))
```

It works, and by watching `nvidia-smi` I can see first GPUs 0 and 1 get utilized, and finally data is transferred to the client on GPU 2.

TL;DR: Scheduler doesn't seem to need information about GPUs (at least for a simple setup like this), but the client does.

**Comment by quasiben** [comment]
@mrocklin did you say ucp initialization *had* to occur before cuda initialization or did I invent this ?  If so, do you know where this requirement came from ?

**Comment by Akshay-Venkatesh** [comment]
Ben, my comment https://github.com/rapidsai/ucx-py/pull/180#issuecomment-530957458 doesn't mean that UCP has to be initialized before cuda initialization. It only means that you can't use `UCX_MEMTYPE_CACHE ` if you're going to initialize and allocate cuda buffers before UCP initialization. Since `UCX_MEMTYPE_CACHE=n` in most of the experiments you run, cuda can be initialized before UCP initialization as well. Lemme know if there's any further confusion.

**Comment by quasiben** [comment]
Ah!  Thanks for clarifying!

**Comment by quasiben** [comment]
closing in favor of https://github.com/rapidsai/ucx-py/issues/187
